engine/engine.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package engine
+
+import (
+	"os"
+
+	"github.com/rs/zerolog"
+)
+
+// Engine is an engine that is composed of multiple components.
+type Engine struct {
+	log        zerolog.Logger
+	components []*component
+
+	interrupt chan os.Signal
+	notify    chan error
+}
+
+// New creates a new engine.
+func New(log zerolog.Logger, name string, interrupt chan os.Signal) *Engine {
+	e := Engine{
+		log:       log.With().Str("engine", name).Logger(),
+		interrupt: interrupt,
+	}
+
+	return &e
+}
+
+// Component registers a new component for the engine. Components will be shut down
+// in the same order as the one in which they were registered.
+func (e *Engine) Component(name string, run func() error, stop func()) *Engine {
+	c := component{
+		log:  e.log.With().Str("component", name).Logger(),
+		run:  run,
+		stop: stop,
+	}
+
+	e.components = append(e.components, &c)
+
+	return e
+}
+
+// Run launches the engine components and waits for them to either finish successfully,
+// fail, or for an external signal to shut the engine down.
+func (e *Engine) Run() error {
+	e.notify = make(chan error, len(e.components))
+	for _, component := range e.components {
+		go component.Run(e.notify)
+	}
+
+	// Here, we are waiting for a signal, or for one of the components to fail
+	// or finish. In both cases, we proceed to shut down everything, while also
+	// entering a goroutine that allows us to force shut down by sending
+	// another signal.
+	select {
+	case <-e.interrupt:
+		e.log.Info().Msg("engine stopping")
+		e.stop()
+	case err := <-e.notify:
+		if err != nil {
+			e.log.Error().Err(err).Msg("engine failed")
```

Comment:
It's not the engine that failed, but the component. I don't like that we are logging the error twice, once for the component and once for the engine. IMO, we should use `e.notify` _only_ when a component fails, otherwise we can use the `e.interrupt` to break out here in _all_ cases. So when we receive something on `e.notify` we should just say `engine failed due to component failure`, and the error is already logged in the component.